Migrate camunda builtins from feel-editor to seperate repo

Buckwich opened the pull request:
Related to https://github.com/camunda/camunda-modeler/issues/3983

### Proposed Changes

Migration of camundaBuiltins into separate repo.

As this PR is just exporting a list of variables, nothing much to try out (or even test) here. But you can checkout https://github.com/camunda/bpmnlint-plugin-camunda-compat/pull/206 to see how it integrates

### Checklist

To ensure you provided everything we need to look at your PR:

* [x] __Brief textual description__ of the changes present
* [ ] __Visual demo__ attached
* [ ] __Steps to try out__ present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [x] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

<!--
Thanks for creating this pull request! ❤️
-->

barmac commented on src/builtins/util.js:
```diff
@@ -0,0 +1,40 @@
+    description
+  } = builtin;
+
+  const match = name.match(/^([\w\s]+)\((.*)\)$/);
```
Can we do this at the compile time?

Buckwich replied:
yes that would have also been my next steps, but in this PR i just wanted to get it out of the other repo to make it easier to review changes

barmac commented on karma.conf.js:
We could have an `npm start` setup with a feel editor though (so example config).

Buckwich replied:
saw that comment too late, removed karma already

Buckwich commented:
@barmac I have adjusted PR with those changes:
- focused on only including the builtins (eg no contextBuilder)
- removed testing as there is not really anything to test anymore
- updated some deps (eslint 9, npm-run-all2)
- parsing the builtins during compile. for domify its still required to map the json (we could move that to feel-editor only)
- I have squashed this branch already
  - but have kept the commits seperatly on https://github.com/camunda/feel-builtins/compare/main...migrate-detailed if you only want to see the latest changes

Buckwich commented on test/.eslintrc:
migrated to eslint 9

Buckwich commented on src/language/index.js:
removed from here, keeping it lean

Buckwich commented on test/language/index.spec.js:
```diff
@@ -0,0 +1,13 @@
+  it('should work',function() {
+
+    const context = (createContext(camundaBuiltins));
+    expect(Object.keys(context).lentth).to.be.equal(context.length);
```
tests removed

Buckwich commented on package.json:
```diff
@@ -3,7 +3,53 @@
+    "pull:camunda-docs": "git -C camunda-docs pull || git clone git@github.com:camunda/camunda-docs.git camunda-docs",
+    "compile:builtins": "node scripts/compileBuiltins.js",
+    "start": "cross-env SINGLE_START=true npm run dev",
+    "start:camunda": "cross-env SINGLE_START=camunda npm run dev",
```
cleaned up the scripts

nikku commented on .eslintrc:
```diff
@@ -0,0 +1,17 @@
+    "overrides": [
+        {
+            "files": [
+                "karma.conf.js",
```
We don't have Karma anymore, remove that file.

nikku commented on .eslintrc:
```diff
@@ -0,0 +1,17 @@
+{
```
Alternatively: Remove this file. We don't need it anymore.

nikku commented on src/index.js:
```diff
@@ -0,0 +1,26 @@
+import { domify } from 'min-dom';
```
Does it make sense to slice things like this? What is the value of domifying `builtin.info` here, when we could just expose the JSON here, and keep domification in the consuming library.

nikku replied:
(We'd also get rid of rollup in this library, in the process).

nikku replied:
What would be the best option, considering the different usage scenarios we anticipate:

* Usage in `feel-editor`
* Usage in `?` (the other place that we build it for)

In other words: What is the positive or negative impact on either. I'd not care too much about 2 or 3, because UX wise for the user, these are the same.

Questions are:

* Shall we expose JSON or JS
* Shall this library depend on the DOM, or should it work without it (`min-dom` is a DOM utility)

Everything else follows, I believe.

Buckwich replied:
domify is only needed for feel-editor, so lets keep it out. 

exposing js would help with typings, as its directly available to the consumer without parsing

so i'll go with 3

barmac replied:
I'd look up the moddle schema libraries where we expose JSON only: https://github.com/camunda/zeebe-bpmn-moddle

nikku commented on package.json:
```diff
@@ -3,7 +3,27 @@
   "description": "Collection of builtin Camunda extensions for FEEL",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "all": "run-s lint build",
```
Unless there is a clear reason, make this an ES module (`type: "module"`).

Buckwich commented on README.md:
````diff
@@ -1,3 +1,58 @@
+Build the project:
+
+```sh
+npm run build
````
I'm open to change, but I prefer individual `code` blocks per command. It makes it easier to only copy the command you actully want 
<img width="635" height="391" alt="image" src="https://github.com/user-attachments/assets/413b21de-48ed-4213-94cb-242452794c72" />

nikku replied:
Let's be short and consistent ([diagram-js](https://github.com/bpmn-io/diagram-js), [bpmn-moddle](https://github.com/bpmn-io/bpmn-moddle), [bpmn-js](https://github.com/bpmn-io/bpmn-js), [Camunda Modeler](https://github.com/camunda/camunda-modeler), [bpmn-js-element-templates](https://github.com/bpmn-io/bpmn-js-element-templates)).

Developer getting started is just one part of the purpose of the README. The major part is usage, this is where we should aim for a great UX.

Buckwich replied:
as we no longer built i've only kept the install and update commands, which should be consistent with bpmn-js

Buckwich commented on package.json:
```diff
@@ -3,7 +3,27 @@
+    "build": "rollup -c --bundleConfigAsCjs",
+    "build:watch": "npm run build -- -w",
+    "prepare": "run-s build",
+    "builtins": "run-s pull:camunda-docs compile:builtins",
```
as we no longer build i have removed the all script altogether

Buckwich replied:
BTW: the reason I dont update in all:
<img width="471" height="149" alt="image" src="https://github.com/user-attachments/assets/85a92199-4b81-4e0f-b88b-3c8a80c28785" />

I know I can adjust the permissions of the token but i would like to keep the CI action permission minimal

nikku commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,25 @@
+      - name: Setup project
+        uses: bpmn-io/actions/setup@latest
+      - name: Lint
+        run: npm run lint
```
On CI, similar to local development, the script to execute is typically `npm run all`. If CI and local `all` script differs, call the CI one `ci`.

```suggestion
        run: npm run all
```

nikku commented on README.md:
```diff
@@ -1,3 +1,53 @@
+* [Changelog](./CHANGELOG.md)
+* [Issues](https://github.com/camunda/feel-builtins/issues)
+
+## Setup and update builtins 
```
```suggestion
## Build and Run
```

nikku commented on src/camundaBuiltins.js:
```diff
@@ -0,0 +1,1718 @@
+export const camundaBuiltins = [
```
I'd consider to rename this to `camunda.js`.

Given the context it is clear to me that this is built-ins.

nikku replied:
Also, given that this is not transpiled anymore, move to `lib` folder.

Buckwich replied:
regarding `lib` i was looking for other repos where we don't build, I only found https://github.com/bpmn-io/bpmn-js-executable-fix and that used src as well

nikku replied:
> regarding camunda vs camundaBuiltins. I like to name the file the same as the main content. For that consider the consumer:

Yes. Would work, too and make the type explicit.

nikku commented on scripts/compileBuiltins.mjs:
```diff
@@ -0,0 +1,45 @@
+import { glob } from 'glob';
```
Rename to `.js`, `m` is not needed inside of an ES module.

nikku commented on README.md:
```diff
@@ -1,3 +1,53 @@
+
+## Usage
+
+You can use the builtins in your FEEL editor by importing them from this package. Here's an example of how to integrate them into a FEEL editor instance:
```
Link `feel-editor`, consider integration with linting.

```suggestion
In your [FEEL editor](https://github.com/bpmn-io/feel-editor) you can use these builtins to establish the Camunda context:
```

Consider: Should this example also feature the Camunda dialect? I think that would make sense.

Consider: Should we also feature integration with feel-lint? I think that would make sense.

nikku commented on scripts/compileBuiltins.mjs:
```diff
@@ -0,0 +1,45 @@
+import { glob } from 'glob';
```
Nit-pick, cf. [bpmn-js](https://github.com/bpmn-io/bpmn-js) and others: Maintain auxiliary scripts in `tasks`.

Buckwich replied:
i actually had problems importing utils without mjs. i can look into that

nikku replied:
Good, let's fix / align that.

Buckwich replied:
once feel-editor is migrated that will not have scripts anyway (cf https://github.com/bpmn-io/feel-editor/compare/3983_migrate-builtins)

nikku commented on package.json:
```diff
@@ -2,14 +2,25 @@
   ],
-  "main": "dist/index.js",
-  "module": "dist/index.es.js",
+  "main": "src/index.js",
```
Let's add a `"sideEffects": false` annotation, to allow tree-shaking in the future - https://github.com/bpmn-io/bpmn-js/blob/develop/package.json#L62.

nikku commented on package.json:
```diff
@@ -2,14 +2,25 @@
   ],
-  "main": "dist/index.js",
-  "module": "dist/index.es.js",
+  "main": "src/index.js",
```
In fact we [don't ship main anymore](https://github.com/bpmn-io/min-dom/blob/main/package.json) on an ES module.

Buckwich commented on README.md:
````diff
@@ -1,3 +1,53 @@
+```
+
+Update the builtins
+```sh
+# pull camunda-docs and extract builtins
+npm run update:builtins
+```
````
i removed testing due to this comment: https://github.com/camunda/feel-builtins/pull/1#discussion_r2205225939

Buckwich replied:
which runner should i use,?
I was thinking of just mocha (as in https://github.com/camunda/zeebe-bpmn-moddle/blob/main/package.json)

barmac replied:
browser needed -> karma + mocha
no browser -> mocha

Buckwich replied:
added some basic tests

nikku commented on src/index.js:
```diff
@@ -0,0 +1 @@
+export * from './camundaBuiltins.js';
```
Works for me, and `camundaBuiltins` works for me, too. I'd just like us to be explicit what we export, under what name here.

From the consumer perspective, this is fine:

```
import { camunda as camundaBuiltins } from '@camunda/feel-builtins';
```

This is fine, too:


```
import { camundaBuiltins } from '@camunda/feel-builtins';
```


Regarding [variable naming](https://sentry.io/answers/naming-variables-in-javascript/#:~:text=As%20JavaScript%20variables%20do%20not%20have%20set%20types%2C%20it%20can%20be%20useful%20to%20include%20an%20indication%20of%20the%20type%20in%20the%20name%20(e.g.%20orderNumber%20is%20obviously%20a%20numeric%20ID%2C%20whereas%20order%20could%20be%20an%20object%2C%20a%20string%20or%20anything%20else)): I see the point. What is important in that consideration is to define `Builtin` as a type, too. Do we intend to ship type definitions with this library? I think it would make sense.

Buckwich commented on lib/camundaBuiltins.js:
```diff
@@ -0,0 +1,1726 @@
+    "type": "function",
+    "params": [
+      {
+        "name": "list， size"
```
lovely, needs to be fixed at source https://github.com/camunda/camunda-docs/blob/2708c47ff1500416bf6faf283a9a604db35ecacc/docs/components/modeler/feel/builtin-functions/feel-built-in-functions-list.md?plain=1#L625

nikku commented on README.md:
```diff
@@ -1,3 +1,82 @@
+import { lintExpression } from '@bpmn-io/feel-lint';
+
+function createContext(builtins) {
+  return builtins.reduce((context, builtin) => {
```
Yea, let's do that, and it is obvious from the README. A great benefit of documentation driven develoment.

Buckwich commented:
Tested https://github.com/camunda/feel-builtins/pull/3 for comments in PRs. Will go ahead and release now
